src/LazyArray.h

```diff
@@ -0,0 +1,127 @@
+
+#pragma once
+
+/** A dynamic array that defers allocation to the first modifying access.
+Const references from the array before allocation will all be to the same default constructed value.
+It is therefore important that default constructed values are indistinguishable from each other. */
+template <typename T>
+class cLazyArray
+{
+	static_assert(!(std::is_reference<T>::value || std::is_array<T>::value),
+		"cLazyArray<T>: T must be a value type");
+	static_assert(std::is_default_constructible<T>::value,
+		"cLazyArray<T>: T must be default constructible");
+public:
+	using value_type = T;
+	using pointer = T *;
+	using const_pointer = const T *;
+	using reference = T &;
+	using const_reference = const T &;
+	using size_type = int;
+	using iterator = pointer;
+	using const_iterator = const_pointer;
+	
+	cLazyArray(size_type a_Size) NOEXCEPT:
+		m_Size{ a_Size }
+	{
+		ASSERT(a_Size > 0);
+	}
+
+	cLazyArray(const cLazyArray & a_Other):
+		m_Size{ a_Other.m_Size }
+	{
+		if (a_Other.IsStorageAllocated())
+		{
+			std::copy(a_Other.begin(), a_Other.end(), begin());
+		}
+	}
+
+	cLazyArray(cLazyArray && a_Other) NOEXCEPT:
+		m_Array{ std::move(a_Other.m_Array) },
+		m_Size{ a_Other.m_Size }
+	{
+	}
+
+	cLazyArray & operator = (const cLazyArray & a_Other)
+	{
+		cLazyArray(a_Other).swap(*this);
+		return *this;
+	}
+
+	cLazyArray & operator = (cLazyArray && a_Other) NOEXCEPT
+	{
+		m_Array = std::move(a_Other.m_Array);
+		m_Size = a_Other.m_Size;
+		return *this;
+	}
+
+	T & operator [] (size_type a_Idx)
+	{
+		ASSERT((0 <= a_Idx) && (a_Idx < m_Size));
+		return data()[a_Idx];
+	}
+
+	const T & operator [] (size_type a_Idx) const
+	{
+		return GetAt(a_Idx);
+	}
+
+	// STL style interface
+
+	const T * cbegin() const { return data(); }
+	T *        begin()       { return data(); }
+	const T *  begin() const { return cbegin(); }
+
+	const T * cend() const { return data() + m_Size; }
+	T *        end()       { return data() + m_Size; }
+	const T *  end() const { return cend(); }
+
+	size_type size() const NOEXCEPT { return m_Size; }
+	
+	const T * data() const
+	{
+		if (m_Array == nullptr)
+		{
+			m_Array.reset(new T[m_Size]);
+		}
+		return m_Array.get();
+	}
+
+	T * data()
+	{
+		auto const_this = const_cast<const cLazyArray *>(this);
+		return const_cast<T *>(const_this->data());
```

Comment:
That makes sense but what if I had `cLazyArray<const int>` (*): `m_Array::element_type` would be `const int` but the `const_cast` would still be okay as it wouldn't actually remove the constness.

(*) Which admittedly is a pretty silly thing to do.

Reply:
`cLazyArray<const int>` is not the problem, since this would invalidate the template at `const T & operator[]` (double const). I meant the problem if somebody should change `std::unique_ptr<T>` to `std::unique_ptr<const T>` (*), because he figured out, that we only access it in const functions.

(*) Your * and the unique_ptr decreases the chance of anybody doing so. But better make the code DAU secure, than searching hours for the error 😼 .